src/Client/ClientApplicationBase.cpp

```diff
@@ -0,0 +1,507 @@
+#include <Client/ClientApplicationBase.h>
+
+#include <base/argsToConfig.h>
+#include <base/safeExit.h>
+#include <Core/BaseSettingsProgramOptions.h>
+#include <Common/clearPasswordFromCommandLine.h>
+#include <Common/TerminalSize.h>
+#include <Common/Exception.h>
+
+#include <Common/config_version.h>
+#include "config.h"
+
+#include <unordered_set>
+#include <string>
+#include <boost/algorithm/string/case_conv.hpp>
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+using namespace std::literals;
+
+namespace CurrentMetrics
+{
+    extern const Metric MemoryTracking;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int CANNOT_SET_SIGNAL_HANDLER;
+    extern const int UNRECOGNIZED_ARGUMENTS;
+}
+
+static ClientInfo::QueryKind parseQueryKind(const String & query_kind)
+{
+    if (query_kind == "initial_query")
+        return ClientInfo::QueryKind::INITIAL_QUERY;
+    if (query_kind == "secondary_query")
+        return ClientInfo::QueryKind::SECONDARY_QUERY;
+    if (query_kind == "no_query")
+        return ClientInfo::QueryKind::NO_QUERY;
+    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unknown query kind {}", query_kind);
+}
+
+/// This signal handler is set only for SIGINT and SIGQUIT.
+void interruptSignalHandler(int signum)
+{
+    if (ClientApplicationBase::getInstance().tryStopQuery())
```

Comment:
This seems to be the cause of the https://github.com/ClickHouse/ClickHouse/issues/66589 crash